crates/net/network/src/manager.rs

```diff
@@ -567,202 +567,236 @@ where
+                                &error,
+                            );
+
+                            this.metrics
+                                .outgoing_connections
+                                .set(this.swarm.state().peers().num_outbound_connections() as f64);
+                        }
+                        SwarmEvent::BadMessage { peer_id } => {
+                            this.swarm.state_mut().peers_mut().apply_reputation_change(
+                                &peer_id,
+                                ReputationChangeKind::BadMessage,
+                            );
+                            this.metrics.invalid_messages_received.increment(1);
+                        }
+                        SwarmEvent::ProtocolBreach { peer_id } => {
+                            this.swarm.state_mut().peers_mut().apply_reputation_change(
+                                &peer_id,
+                                ReputationChangeKind::BadProtocol,
+                            );
+                        }
                     }
-                    this.metrics.closed_sessions.increment(1);
-                    this.metrics
-                        .outgoing_connections
-                        .set(this.swarm.state().peers().num_outbound_connections() as f64);
-                }
-                SwarmEvent::OutgoingConnectionError { remote_addr, peer_id, error } => {
-                    trace!(
-                        target : "net",
-                        ?remote_addr,
-                        ?peer_id,
-                        ?error,
-                        "Outgoing connection error"
-                    );
-
-                    this.swarm.state_mut().peers_mut().on_outgoing_connection_failure(
-                        &remote_addr,
-                        &peer_id,
-                        &error,
-                    );
-
-                    this.metrics
-                        .outgoing_connections
-                        .set(this.swarm.state().peers().num_outbound_connections() as f64);
-                }
-                SwarmEvent::BadMessage { peer_id } => {
-                    this.swarm
-                        .state_mut()
-                        .peers_mut()
-                        .apply_reputation_change(&peer_id, ReputationChangeKind::BadMessage);
-                    this.metrics.invalid_messages_received.increment(1);
-                }
-                SwarmEvent::ProtocolBreach { peer_id } => {
-                    this.swarm
-                        .state_mut()
-                        .peers_mut()
-                        .apply_reputation_change(&peer_id, ReputationChangeKind::BadProtocol);
                 }
             }
+
+            // ensure we still have enough budget for another iteration
+            budget -= 1;
+            if budget == 0 {
+                // make sure we're woken up again
+                cx.waker().wake_by_ref();
+                break
+            }
```

Comment:
Ah OK makes sense. This has always concerned me with the `loop`-style endless futures, and the blogpost clarified the solution. Thanks.